<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Задания" data-toc-modified-id="Задания-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Задания</a></span></li></ul></li><li><span><a href="#Создание-подключения" data-toc-modified-id="Создание-подключения-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Создание подключения</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#publishers" data-toc-modified-id="publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>publishers</a></span></li><li><span><a href="#ratings" data-toc-modified-id="ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>ratings</a></span></li><li><span><a href="#reviews" data-toc-modified-id="reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>reviews</a></span></li></ul></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Количество-книг,-вышедших-после-1-января-2000-года" data-toc-modified-id="Количество-книг,-вышедших-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Количество книг, вышедших после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средняя-оценка" data-toc-modified-id="Количество-обзоров-и-средняя-оценка-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Количество обзоров и средняя оценка</a></span></li><li><span><a href="#Издательство-с-наибольшим-количеством-книг" data-toc-modified-id="Издательство-с-наибольшим-количеством-книг-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Издательство с наибольшим количеством книг</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книг" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книг-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книг</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей,-которые-поставили-более-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей, которые поставили более 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Описание данных

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Таблица **books** содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors** содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers** содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица **ratings** содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews** содержит данные о пользовательских обзорах на книги:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Создание подключения

Создадим подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
    'pwd': 'Sdf4$2;d-d30pp', # пароль
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
    'port': 6432, # порт подключения
    'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
    db_config['pwd'], 
    db_config['host'], 
    db_config['port'], 
    db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Таблица books

In [12]:
pd.io.sql.read_sql('''
SELECT *
FROM books
''', con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе содержится информация о 1000 книг

### Таблица authors

In [13]:
pd.io.sql.read_sql('''
SELECT *
FROM authors
''', con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

Эти книги написали 636 авторов

### publishers

In [14]:
pd.io.sql.read_sql('''
SELECT *
FROM publishers
''', con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Всего 340 издательств, которые выпустили эти книги

### ratings

In [15]:
pd.io.sql.read_sql('''
SELECT *
FROM ratings
''', con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Книги получили 6,5 тысяч оценок от пользователей

### reviews

In [16]:
pd.io.sql.read_sql('''
SELECT *
FROM reviews
''', con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

И на данные книги написано 2800 отзывов

## Задания

### Количество книг, вышедших после 1 января 2000 года

In [7]:
pd.io.sql.read_sql('''
SELECT COUNT(*)
FROM books
WHERE publication_date >= '2000-01-01'
''', con = engine)

count
0    821

В период с 01 января 2000 года вышла 821 книга

### Количество обзоров и средняя оценка 

Посчитаем среднюю оценку и количество отзывов для каждой книги

In [17]:
pd.io.sql.read_sql('''
SELECT b.title,
       AVG(rt.rating),
       COUNT(rw.review_id)
FROM books AS b
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
LEFT JOIN reviews AS rw ON b.book_id = rw.book_id
GROUP BY b.book_id
ORDER BY AVG(rt.rating) DESC,
         COUNT(rw.review_id) DESC
''', con = engine)

title   avg  count
0                        A Dirty Job (Grim Reaper  #1)  5.00     16
1              School's Out—Forever (Maximum Ride  #2)  5.00     12
2         Moneyball: The Art of Winning an Unfair Game  5.00      9
3    In the Hand of the Goddess (Song of the Liones...  5.00      6
4                             Tai-Pan (Asian Saga  #2)  5.00      4
..                                                 ...   ...    ...
995  The World Is Flat: A Brief History of the Twen...  2.25     12
996                                      Drowning Ruth  2.00      9
997                                              Junky  2.00      4
998                  His Excellency: George Washington  2.00      4
999                               Harvesting the Heart  1.50      4

[1000 rows x 3 columns]

### Издательство с наибольшим количеством книг

Определим издательство, которое выпустило больше всего книг. Примем условие, что книга содержит более 50 страниц, чтобы отбросить брошюры

In [9]:
pd.io.sql.read_sql('''
SELECT p.publisher,
       COUNT(b.book_id)
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY COUNT(b.book_id) DESC
LIMIT 1
''', con = engine)

publisher  count
0  Penguin Books     42

### Автор с самой высокой средней оценкой книг

Найдём автора, который имеет самую высокую среднюю оценку за все выпущенные им книги.

In [10]:
pd.io.sql.read_sql('''
SELECT a.author,
       AVG(rt.rating) 
FROM books AS b
LEFT JOIN authors AS a ON b.author_id = a.author_id
LEFT JOIN ratings AS rt ON b.book_id = rt.book_id
GROUP BY a.author
HAVING COUNT(rt.rating) > 50
ORDER BY AVG(rt.rating) DESC
LIMIT 1
''', con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

### Среднее количество обзоров от пользователей, которые поставили более 50 оценок

Найдём среднее количество обзоров от пользователей, которые поставили более пятидесяти оценок. Для этого создадим две временные таблицы: в первой отберём тех, кто поставил более 50 оценок, а во второй - посчитаем, сколько отзывов оставляют пользователи.

In [22]:
pd.io.sql.read_sql('''
WITH 
    user_rating AS (
        SELECT username,
               COUNT(rating_id) AS rate
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    ), 
    user_review AS (
        SELECT username,
               COUNT(review_id) AS rew
        FROM reviews
        GROUP BY username
    )
SELECT AVG(rew)
FROM user_rating LEFT JOIN user_review ON user_rating.username = user_review.username
''', con = engine)

avg
0  24.333333

## Вывод

1. В нашей базе 1000 книг, 636 записей об авторах и 340 издательств
2. Пользователи оставили 6.5 тысяч оценок книгам и 2.8 тысяч отзывов
3. После 1 января 2000 года вышла 821 книга
4. Самая высоко-оценённая книга имеет рейтинг 5 и 16 отзывов
5. Больше всего книг выпустило издание Penguin Books. Они выпустили 42 книги
6. Самая высокая средняя оценка за книги у автора - у создательницы серии книг про Гарри Поттера Джоан Роулинг
7. Пользователи, которые поставили более 50 оценок книгам, оставляют в сренем 24 отзыва.